cli/remote_build.go

```diff
@@ -888,45 +905,89 @@ func (cmd *RemoteCmd) submitBuild(project *Project, tagMap map[string]string) er
 	log.Infof("Submission finished at %s, taking %s", endTime.Format(TIME_FORMAT), submitTime.Truncate(time.Millisecond))
 
 	startTime = time.Now()
-	var remoteProgress *Progress
-	remoteErrored := func() {
-		if remoteProgress != nil {
-			remoteProgress.Fail()
-		}
-	}
 	if log.CheckIfTerminal() {
 		remoteProgress = NewProgressSpinner("Setting up remote build")
 		remoteProgress.Start()
 	}
 
 	if strings.HasPrefix(url, "ws:") || strings.HasPrefix(url, "wss:") {
+		recentReconnect := false
+		reconnectCount := 0
+	CONN:
 		conn, _, _, err := ws.DefaultDialer.Dial(context.Background(), url)
+
+		finish := make(chan struct{})
+
 		if err != nil {
 			remoteErrored()
 			return fmt.Errorf("Cannot connect: %v", err)
 		} else {
 
+			// TODO maybe make Dispatcher give better diagnostics somehow
+			go func() {
+				for {
+					select {
+					case <-finish:
+						return
+					case <-time.After(5 * time.Second):
+						if err := wsutil.WriteClientMessage(conn, ws.OpPing, []byte("remotebuild ping")); err != nil {
+							log.Errorf("Cannot send ping: %v", err)
+						}
+					}
+				}
+			}()
+
 			defer func() {
 				if err = conn.Close(); err != nil {
 					log.Debugf("Cannot close: %v", err)
 				}
 			}()
 
 			buildSuccess := false
+			buildFailed := false
 			buildSetupFinished := false
 			for {
 				msg, control, err := wsutil.ReadServerData(conn)
 				if err != nil {
-					if err != io.EOF {
-						log.Tracef("Unstable connection: %v", err)
-					} else {
+					if err == io.EOF {
```

Comment:
In code that targets Go 1.13 or above, prefer using [`errors.Is`](https://pkg.go.dev/errors?tab=doc#Is):

```suggestion
					if errors.Is(err, io.EOF) {
```